Bluetooth: Mesh: More model publication and related fixes/cleanups

jhedberg opened the pull request.

Commit:
Bluetooth: Mesh: Fix non-periodic Model Publication

The only generally available model supporting publication that's
convenient to be used for testing is the Health Server Model.
Unfortunately since this model supports period publication, the
non-periodic side got less attention and had some bugs.

The first thing that needs to be done is to verify that the period
returned by bt_mesh_model_pub_period_get() is positive. If it's zero
then no periodic publication should take place.

Another thing that this patch cleans up is the naming of the callback
used for periodic publishing. There's no need do require the callback
to call bt_mesh_model_publish() since this must happen no matter what,
so instead rename the callback from 'func' to 'update' and have the
access layer call bt_mesh_model_publish() if the callback was
successful.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
Bluetooth: samples/mesh_demo: Remove unused model publication

This app isn't (at least yet) using the model publication for
anything, and in fact this could cause trouble due to missing
publication net_buf_simple buffer.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
Bluetooth: tests/mesh: Fix model publication

The app was declaring model publication contexts but missing buffers
for the publication messages.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
Bluetooth: Mesh: Provide documentation for the publication API

Document all members of the bt_mesh_model_pub struct.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
Bluetooth: Mesh: Fix accuracy of Publish Period

After the Publish Retransmit state was introduced the Publish Period
measurement would begin once the previous Publish message has finished
transmitting. This will however cause inaccurate periods, which is
particularly an issue with the PTS that expects accuracy of less than
0.5 seconds (apparently).

Since the publication timer is also used for the retransmissions we
can't simultaneously use if for the period as well. Therefore, we
introduce a new variable called period_start which makes a note of
when the period was supposed to start, and then once all
retransmissoins are done initializes the timer with the send duration
taken into account.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
Bluetooth: tests/mesh_shell: Fix main stack size for qemu

The qemu main() seems to require more stack than the arm equivalent.
Increase the stack size to 320 to avoid crashes.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
Bluetooth: Mesh: remove redundant 32-bit wrap-around code

The separate checking for "now < reftime" is unnecessary, since the
integer over/under-flow for unsigned 32-bit values resulting from
subtraction will give the right delta even if 'now' is less than the
reference.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

carlescufi commented on include/bluetooth/mesh/access.h:
```diff
@@ -256,27 +256,48 @@ struct bt_mesh_model_op {
+	 *  The application is expected to initialize this with
+	 *  a valid net_buf_simple pointer, with the help of e.g.
+	 *  the NET_BUF_SIMPLE() macro. The buffer must contain
+	 *  a valid publication messages before calling the
```
remove either "a" or the "s" in "messages"

carlescufi replied:
Use the term "publication buffer" for consistency with the callback doc, or change it in both places to make sure users know that the callback is tied to this buffer.

carlescufi commented on include/bluetooth/mesh/access.h:
```diff
@@ -256,27 +256,48 @@ struct bt_mesh_model_op {
+	 *  When set to NULL, the model is assumed not to support
+	 *  periodic publishing. When set to non-NULL the callback
+	 *  will be called periodically and is expected to update
+	 *  the publication buffer with a valid publication message.
```
Maybe refer explicitly here to bt_mesh_model_pub :: msg

jhedberg replied:
Fixed now, however due to conflicts I ended up fixing this as a new patch rather than amending the existing one. I hope that's ok.

Commit:
Bluetooth: Mesh: Improve model publication documentation

Make it clear which members of bt_mesh_model_pub the documentation is
referencing.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
Bluetooth: Mesh: Fix dereferencing mod->pub too early

The bt_mesh_model_publish() is supposed to return a "not supported"
error if the publish context doesn't exist. Fixing the premature
dereferencing also fixes coverity warnings.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
Bluetooth: Mesh: Fix coverity warning of unchecked return

Most places check the return value of bt_mesh_model_send(), so check
for it in all places.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
Bluetooth: Mesh: Friend: Fix poll timeout type

The Poll Timeout needs to be at least 24 bits, so u16_t doesn't
suffice and will potentially result in truncation. Use u32_t, thereby
also fixing a coverity warning.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

jhedberg commented:
@carlescufi I've done the documentation changes - could you check if you're ok with the PR now?

Vudentz reviewed (approved):
LGTM

carlescufi reviewed (approved):
Comments addressed

jhedberg commented:
@rymanluk @michal-narajowski FYI
